File: bitcamp-web-04/src/main/java/bitcamp/pms/dao/MemberDao.java
package bitcamp.pms.dao;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

import bitcamp.pms.domain.Member;

public class MemberDao {
    
    
/*    static String jdbcUrl = "jdbc:mysql://13.124.153.245:3306/studydb"; //스태틱은 같은 스태틱 멤버끼리...
    //같은 스태틱 멤버여야 한다그래서 static으로 선언
    static String username = "study";
    static String password = "1111";*/

    //memberdao 로딩될때 한번만 하도록 하겠음 Class.
    static {
        try {
            Class.forName("com.mysql.jdbc.Driver");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    String jdbcUrl; //DBMS 따로따로 관리하기 위해 인스턴스 변수로 만듬
    String username;
    String password;
    
    
    
    public MemberDao(String jdbcUrl, String username, String password) {
        this.jdbcUrl = jdbcUrl;
        this.username = username;
        this.password = password;
    } //MemberDao를 따로따로 관리할 수 있음
    
    public List<Member> selectList() throws Exception{ //위 인스턴스 변수를 만들고 나서 static을 없앰
      //Class.forName("com.mysql.jdbc.Driver");*/ //클래스 로딩
      //그리고 전체 Class.forname("com.mysql.jdbc.Driver") 삭제
        try (

                Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
                PreparedStatement stmt = con.prepareStatement(
                    "select mid, email from pms2_member");
                ResultSet rs = stmt.executeQuery();) {
                
                ArrayList<Member> list = new ArrayList<>();
                while (rs.next()) {
                    Member member = new Member();
                    member.setId(rs.getString("mid")); 
                    member.setEmail(rs.getString("email"));
                    list.add(member);
                }
                return list;

            }
    }
    
    public Member selectOne(String id) throws Exception{

        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "select mid,email from pms2_member where mid=?");) { 
        
            stmt.setString(1, id);

            try (ResultSet rs = stmt.executeQuery();) {
                if (!rs.next()) { 
                       return null;
                }
                Member member = new Member();
                member.setId(rs.getString("mid"));
                member.setEmail(rs.getString("email"));
                return member;
            }
            //out.println("<p>유효하지 않은 멤버 아이디입니다.</p>")
        }
    }
    
    public int update(Member member) throws Exception{

        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "update pms2_member set email=?, pwd=password(?) where mid=?");) {
            
            stmt.setString(1, member.getEmail());
            stmt.setString(2, member.getPassword());
            stmt.setString(3, member.getId());
            
            return stmt.executeUpdate();
        }
    }
    
    public int delete(String id) throws Exception{

        try (
            Connection con = DriverManager.getConnection(
                    jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "delete from pms2_member where mid=?");) {
            
            stmt.setString(1, id);

            return stmt.executeUpdate();
        }
    }
    
    public void insert(Member member) throws Exception{
        //memberDao.insert(member);

          try (
              Connection con = DriverManager.getConnection(
                      jdbcUrl, username, password);
              PreparedStatement stmt = con.prepareStatement(
                  "insert into pms2_member(mid,email,pwd) values(?,?,password(?))");) {
              
              stmt.setString(1, member.getId());
              stmt.setString(2, member.getEmail());
              stmt.setString(3, member.getPassword());
          
              stmt.executeUpdate();

          }
      }

}

File: bitcamp-nodejs/src/step05/memberdao.js
// 주제: DAO 모듈 만들기

//pool 주입 받기
var pool;

exports.connectiontPool = (connectionPool) => {
    pool = connectionPool;
};


/*
 * 비동기는 동기처럼 위에서 아래로 흐르는게 아니다. 갑자기 중간에서 실행될수도 있고 그래서 디버깅이 가장 힘들다.
 */
exports.list = (pageNo = 1, pageSize = 3, handler) => {
    var startIndex = (pageNo - 1) * pageSize;
    
    pool.query('select mid, email from pms2_member limit ?, ?',
            [startIndex, pageSize],
            function(err, results) {
        /*
         * if (err) { res.end('DB 조회 중 예외 발생!') return; }
         * 
         * for (var row of results) { res.write(`${row.email}, ${row.mid}\n`); }
         * res.end();
         */
        
        handler(err, results); // 작업이 끝나면 이 handler에게 err나 results를 전달하는 것
    });
}

exports.add = (data, handler) => {
    pool.query(
            'insert into pms2_member(mid,email,pwd)\
            values(?, ?, password(?))',
        [data.id,
         data.email,
         data.password],
        function(err, result) {
            handler(err, result);
    });
}

exports.update = (data, handler) => {
    pool.query(
            'update pms2_member set\
             email=?,\
             pwd=?\
             where mid=?',
        [data.email,
         data.password,
         data.id],
        function(err, result) {
             handler(err, result);
    });
}

exports.remove = (data, handler) =>{
    pool.query('delete from pms2_member where mid=?',
            [data.id],
            function(err, result) {
                handler(err, result);
        });
}